File: C#/0166. Fraction to Recurring Decimal.cs
public class Solution {
    public string FractionToDecimal(int numerator, int denominator) {
        long num=numerator;
        long den=denominator;
        
        if(num==0)
            return "0";
        if(num%den==0)
            return (num/den).ToString();
        
        int flag=num*den<0?-1:1;
        num=Math.Abs(num);
        den=Math.Abs(den);
        
        string rep= flag==1?"":"-";
        rep+=(num/den).ToString()+".";
        
        num=num%den;
        var numsDict=new Dictionary<long,int>();
        string str="";
        int i=0;
        while(!numsDict.ContainsKey(num) && num!=0){
            numsDict[num]=i;
            i++;
            num=num*10;
            str+=(num/den).ToString();
            num=num%den;
        }
        
        if(num==0)
            rep+=str;
        else{
            int k=numsDict[num];
            for(int j =0;j<k;j++){
                rep+=str[j];
            }
            rep+='(';
            for(int j=k;j<i;j++){
                rep+=str[j];
            }
            rep+=')';
            }
        return rep;
        
    }
}

File: archives/leetcode2/second/0166. Fraction to Recurring Decimal.py
class Solution:
    def fractionToDecimal(self, numerator: 'int', denominator: 'int') -> 'str':
        
        if numerator%denominator==0:
            return str(numerator//denominator)
        
        flag=''
        if numerator*denominator<0:
            flag='-'
        numerator,denominator=abs(numerator),abs(denominator)
        
        k=numerator//denominator
        numerator=numerator%denominator
        dictdenominator={}
        rep=''
        cnt=0
        while numerator!=0:
            if numerator not in dictdenominator:
                dictdenominator[numerator]=cnt
                cnt+=1
                rep+=str(numerator*10//denominator)
                numerator=numerator*10%denominator
            else:
                m=dictdenominator[numerator]
                if m==0:
                    rep='('+rep+')'
                else:
                    rep=rep[:m]+'('+rep[m:]+')'
                break
        return flag+str(k)+'.'+rep
